### ISSUE TITLE
crop an image

### ISSUE THREAD

**Issue by brunogml** [openseadragon/openseadragon#2527]
Hi, 
I have an image 1000x1000 px of which I want to display only the cropped part from coord. (100,100) to coord. (500,500).

The complete image is displayed with the following script:

```
<script src="seadragon/openseadragon.min.js"></script>
<script type="text/javascript">
    var viewer = OpenSeadragon({
	 maxZoomLevel: 5.0,
        id: "openseadragon1",
        prefixUrl: "seadragon/images/",
	sequenceMode: true,
        tileSources:  [ '.$json.' ]
   });
</script>';
```
$json is the manifest of the image

Which command should I add to crop this image?
Thanks

**Comment by iangilman** [comment]
You want `setClip`:

https://openseadragon.github.io/docs/OpenSeadragon.TiledImage.html#setClip

For example:

```
viewer.addHandler('open', () => {
    viewer.world.getItemAt(0).setClip(new OpenSeadragon.Rect(100, 100, 400, 400));
});
```

**Comment by brunogml** [comment]
Many thanks!

**Comment by iangilman** [comment]
You can do that like so:

```
viewer.addHandler('open', () => {
    const rect = new OpenSeadragon.Rect(100, 100, 400, 400);
    viewer.world.getItemAt(0).setClip(rect);
    viewer.viewport.fitBounds(rect, true);
});
```

That'll automatically zoom you to that location. However it won't reset the home position, so if the user clicks the "home" button it'll still zoom them back out all the way.

You might try `viewer.viewport.setHomeBounds(rect)`. It's deprecated but I think it should still work. Maybe we need to reinstate it, because it seems like the right choice for this situation.

**Comment by Whao510** [comment]
How do I get a cropped image instead of cropping it to affect the original image

**Comment by iangilman** [comment]
OSD doesn't have a feature for that. It might be interesting to explore adding that feature; I imagine it'll take some doing.

Meanwhile, can you crop the image on the server before sending it to OSD?

**Comment by pearcetm** [comment]
> How do I get a cropped image instead of cropping it to affect the original image


> > OSD doesn't have a feature for that. It might be interesting to explore adding that feature; I imagine it'll take some doing.
> > 
> > Meanwhile, can you crop the image on the server before sending it to OSD?

Isn't this what `TiledImage.setCroppingPolygons` does? https://openseadragon.github.io/docs/OpenSeadragon.TiledImage.html#setCroppingPolygons

**Comment by iangilman** [comment]
@pearcetm Both `setClip` and `setCroppingPolygons` just hide parts of the image... They don't change the bounds of the image so OSD would treat the image as if it had been truly cropped. I believe that's what @Whao510 is hoping for.

I could see how a "true crop" where the bounds of the image are changed would be nice for working with data sets the developer doesn't have direct control over (or where there's a desire to crop out things nondestructively). I'm not sure how we would go about it, though, as the true bounds of the image are important for tile handling. Also, of course, we would want to maintain backwards compatibility with apps that expect the current behavior. Perhaps some kind of flag on the image that indicates that the `World` should take into account the clip bounds instead of the full bounds?

I figure the main concern here is just having "home" honor the clip bounds, but perhaps there are other implications as well.

**Comment by pearcetm** [comment]
@iangilman I wasn't clear on exactly what @Whao510 was looking for - you might be right that it was specifically about the bounds. I thought maybe they were just asking how to crop an image at all.

One way that bounds could be handled for clipped and cropped images is keeping the true bounds as they are (important for tile positioning, as you say) and keep a separate variable (`visibleBounds` or `clippedBounds`, maybe) which would be used for constraints, home bounds, etc. For unclipped/uncropped images these would be the same as the original bounds, but would be updated by `setClip` and `setCroppingPolygons` to reflect the visible portion of the image. And yes, a flag to enable this behavior makes sense from a backwards compatibility standpoint.

**Comment by iangilman** [comment]
Yeah, I think that would be a good way to go. We'll see if someone wants to pick up the task.

**Comment by brunogml** [comment]
Sorry, but `viewer.viewport.fitBounds(rect, true);` does not work for my application.
[Here](http://www.rkts.org/images.php?id=1%7CDd%7CMWEAP570-5-1%7CIEAP570-5-1-25%7C%27dul+ba%7C8a) a sample with fitbounds and [here](http://www.rkts.org/images.php?id=1%7CCz%7CMWEAP570-4-1%7CIEAP570-4-1-7%7C%27bum%7C4a) without fitbounds.
A click on the "go home" button of OSD re-zooms and re-locates correctly the cropped image.
Thank you for yur help

**Comment by iangilman** [comment]
@brunogml Looks like you're not giving it the correct rectangle for `fitBounds` (since the image is off the screen at the start). Are you using viewport coordinates?

**Comment by brunogml** [comment]
@iangilman I use the rec defined at the beginning of the Handler

```
viewer.addHandler(\'open\', () => {
	const rect = new OpenSeadragon.Rect('.$l1.', '.$l2.', '.$l3.', '.$l4.');
    viewer.world.getItemAt(0).setClip(rect);
    viewer.viewport.fitBounds(rect, true);
	});
```

**Comment by iangilman** [comment]
Oh! My mistake. For `setClip` the rectangle needs to be in image coordinates, but for `fitBounds` it needs to be in viewport coordinates. Here's how you convert:

```
viewer.addHandler(\'open\', () => {
	const imageRect = new OpenSeadragon.Rect('.$l1.', '.$l2.', '.$l3.', '.$l4.');
    const tiledImage = viewer.world.getItemAt(0);
    tiledImage.setClip(imageRect);
    const viewportRect = tiledImage.imageToViewportRectangle(imageRect);
    viewer.viewport.fitBounds(viewportRect, true);
});
```

For more on the different coordinate systems, see: 

https://openseadragon.github.io/examples/viewport-coordinates/

**Comment by brunogml** [comment]
Fantastic! Thank you so much.
